CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+<sup>[[link](#no-and-or-or)]</sup>
+
+  ```Ruby
+  # bad
+  # boolean expression
+  if some_condition and some_other_condition
+    do_something
+  end
+
+  # control flow
+  document.saved? or document.save!
+
+  # good
+  # boolean expression
+  if some_condition && some_other_condition
+    do_something
+  end
+
+  # control flow
+  document.saved? || document.save!
+  ```
+
+* <a name="no-multiline-ternary"></a>
+  Avoid multi-line `?:` (the ternary operator); use `if/unless` instead.
+<sup>[[link](#no-multiline-ternary)]</sup>
+
+* <a name="if-as-a-modifier"></a>
+  Favor modifier `if/unless` usage when you have a single-line body. Another
+  good alternative is the usage of control flow `&&/||`.
+<sup>[[link](#if-as-a-modifier)]</sup>
+
+  ```Ruby
+  # bad
+  if some_condition
+    do_something
+  end
+
+  # good
+  do_something if some_condition
+
+  # another good option
+  some_condition && do_something
+  ```
+
+* <a name="no-multiline-if-modifiers"></a>
+  Avoid modifier `if/unless` usage at the end of a non-trivial multi-line
+  block.
+<sup>[[link](#no-multiline-if-modifiers)]</sup>
+
+  ```Ruby
+  # bad
+  10.times do
+    # multi-line body omitted
+  end if some_condition
+
+  # good
+  if some_condition
+    10.times do
+      # multi-line body omitted
+    end
+  end
+  ```
+
+* <a name="unless-for-negatives"></a>
+
+Avoid usage of `unless` over negative conditions (or control flow), it
+can cause issues when having multi term expressions.
+<sup>[[link](#unless-for-negatives)]</sup>
+
+  ```Ruby
+  # bad
+  do_something unless some_condition
+
+  # good
+  do_something if !some_condition
+
+  # another good option
+  some_condition || do_something
+  ```
+
+* <a name="no-else-with-unless"></a>
+  Do not use `unless` with `else`. Rewrite these with the positive case first.
+<sup>[[link](#no-else-with-unless)]</sup>
+
+  ```Ruby
+  # bad
+  unless success?
+    puts 'failure'
+  else
+    puts 'success'
+  end
+
+  # good
+  if success?
+    puts 'success'
+  else
+    puts 'failure'
+  end
+  ```
+
+* <a name="no-parens-if"></a>
+  Don't use parentheses around the condition of an `if/unless/while/until`.
+<sup>[[link](#no-parens-if)]</sup>
+
+  ```Ruby
+  # bad
+  if (x > 10)
+    # body omitted
+  end
+
+  # good
+  if x > 10
+    # body omitted
+  end
+  ```
+
+Note that there is an exception to this rule, namely [safe assignment in
+condition](#safe-assignment-in-condition).
+
+* <a name="no-multiline-while-do"></a>
+  Do not use `while/until condition do` for multi-line `while/until`.
+<sup>[[link](#no-multiline-while-do)]</sup>
+
+  ```Ruby
+  # bad
+  while x > 5 do
+    # body omitted
+  end
+
+  until x > 5 do
+    # body omitted
+  end
+
+  # good
+  while x > 5
+    # body omitted
+  end
+
+  until x > 5
+    # body omitted
+  end
+```
+
+* <a name="infinite-loop"></a>
+  Use `Kernel#loop` instead of `while/until` when you need an infinite loop.
```

Comment:
just want to add that in critical sections, I would stick with `while true` because using `loop` implies calling a closure on every iteration - again, this is really micro-optimization and not too elegant Ruby-wise but I can live with that :P